Adjust to new JSON ABI format

topolarity opened the pull request:
Expect some more touch-ups on Monday, but this should be the core pieces.

There are some significant changes to how mangling works in here, but hopefully this feels mostly in-spirit with the existing behavior. As a bonus, we now support recursive types! (which require forward declarations / pointer-indirection in C, of course)

topolarity committed:
Adjust to new JSON ABI format

timholy commented:
Thanks! I'll review shortly. As you probably know, there is no "user base" for this package yet, so feel free to "break" anything you think should be improved.

topolarity committed:
Re-factor: Rename `parselog` to `import_abi_info`

Also add a struct definition to be able to conveniently document the
various bits of output data.

topolarity committed:
Tweak C type name sanitizer

This should remove some gnarly-looking underscores.

topolarity committed:
Handle mangled type name collisions

topolarity committed:
Add `sort_declarations!` tests

topolarity committed:
Fix-up `Int` handling for x86

topolarity committed:
Use `struct ...` when referring to forward-declared types

timholy commented on Project.toml:
```diff
@@ -5,10 +5,12 @@ authors = ["Tim Holy <tim.holy@gmail.com> and contributors"]
 
 [deps]
 Graphs = "86223c79-3864-5bf0-83f7-82e725a168b6"
+JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
```
JSON3/StructTypes might automate some of the json->julia parsing, but I am not sure there would be any real gain and this seems fine to me.

timholy commented on src/JuliaLibWrapping.jl:
```diff
@@ -3,10 +3,10 @@ module JuliaLibWrapping
 
-export parselog, wrapper
-export CProject
+export import_abi_info, wrapper
```
This is an improvement in naming, thanks

timholy commented on src/abi_import.jl:
```diff
@@ -0,0 +1,209 @@
+
+struct PrimitiveTypeDesc
+    # TODO: support custom primitive types?
+    name::String
```
Would it make sense to include `bits`?

timholy commented on src/abi_import.jl:
```diff
@@ -0,0 +1,209 @@
+end
+
+function build_type_graph(typedescs::OrderedDict{Int, TypeDesc};
+                          pointer_filter::Function)
```
Any reason not to make `pointer_filter` the first positional argument so that one could use a `do` block? https://docs.julialang.org/en/v1/manual/style-guide/#Write-functions-with-argument-ordering-similar-to-Julia-Base

topolarity replied:
I find that a bit more opaque at the call-site TBH, if the function name does not clearly communicate what it's going to use the function for

timholy commented on src/abi_import.jl:
```diff
@@ -0,0 +1,209 @@
+    # therefore have to use forward-declarations instead of just sorting declarations.
+    recursive_types = BitSet()
+
+    full_type_graph = build_type_graph(typedescs; pointer_filter = (id)->true)
```
```suggestion
    full_type_graph = build_type_graph(typedescs; pointer_filter = Returns(true))
```

timholy commented on src/abi_import.jl:
```diff
@@ -0,0 +1,209 @@
+end
+
+"""
+    entrypoints, typedescs, forward_declarations = parselog(filename::String)
```
```suggestion
    abi_info = import_abi_info(filename::String)
```

timholy commented on src/abi_import.jl:
```diff
@@ -0,0 +1,209 @@
+    forward_declared::BitSet
+    # A vector of exposed functions from the imported ABI
+    entrypoints::Vector{MethodDesc}
+end
```
Would this benefit from having a `Base.show` method? And should the comments be moved into a docstring?

topolarity replied:
Nothing fancy, but this is something:
```julia
ABIInfo(...) object, with 12 types and 3 entrypoints.
  Types:
    ∘ Float32
    ∘ Ptr{Float32}
    ∘ Int64
    ∘ Ptr{CTree{Float64}}
    ∘ Int32
    ∘ CVector{Float32}
    ∘ CVectorPair{Float32}
    ∘ MyTwoVec
    ∘ Ptr{MyTwoVec}
    ∘ CVector{CTree{Float64}}
    ∘ Float64
    ∘ CTree{Float64}
  Entrypoints:
    ∘ tree_size(tree::CTree{Float64})
    ∘ copyto_and_sum(fromto::CVectorPair{Float32})
    ∘ countsame(list::Ptr{MyTwoVec}, n::Int32)
```

timholy commented on test/runtests.jl:
```diff
@@ -10,76 +14,156 @@ function onlymatch(f, collection)
+            @test collect(keys(sorted)) == Int[5,2,1,4,3]
+        else
+            @test false # unexpected forward declarations
+        end
```
Nice!

timholy reviewed (commented):
LGTM. A few things to think about, and then merge when you feel ready.

JeffBezanson commented on src/abi_import.jl:
```diff
@@ -0,0 +1,209 @@
+        FieldDesc[
+            FieldDesc(
+                field["name"],
+                field["type"],
```
`type_id`?

topolarity committed:
Rename `type` -> `type_id`

topolarity committed:
Address various review feedback

 - Add `Base.show(::IO, ::ABIInfo)
 - Add doc-comments for `ABIInfo`
 - Add addl. fields to PrimitiveTypeDesc
 - Update test ABI JSON file to latest output

topolarity reviewed (commented):
Apologies for the significant delay - this should have been done months ago!

I've been assuming this PR needed significant re-work for some reason, but the updates from https://github.com/JuliaLang/julia/pull/59108 ended up very easy. I think I've covered all existing feedback now, and this appears to be working with the ABI export landed upstream.

@timholy Do you remember if there were any other major TODO's from our prior conversation(s)?

JeffBezanson commented:
Ready to merge?

topolarity commented:
Good to go from my POV

timholy commented:
Glad to see this finished! Is this still the home for this work or has it already been integrated elsewhere?
